src/mongo/tree/MongoDatabaseTreeItem.ts

```diff
@@ -125,20 +126,33 @@ export class MongoDatabaseTreeItem extends AzureParentTreeItem<IMongoTreeRoot> {
 		return new MongoCollectionTreeItem(this, newCollection);
 	}
 
-	executeCommandInShell(command: MongoCommand, context: IActionContext): Thenable<string> {
+	private async executeCommandInShell(command: MongoCommand, context: IActionContext): Promise<string> {
 		context.telemetry.properties["executeInShell"] = "true";
-		return this.getShell().then(shell => shell.exec(command.text));
+
+		// CONSIDER: Re-using the shell instead of disposing it each time would allow us to keep state
+		//  (JavaScript variables, etc.), but we would need to deal with concurrent requests, or timed-out
+		//  requests.
+		let shell = await this.createShell();
+		try {
+			await shell.useDatabase(this.databaseName);
+			return await shell.executeScript(command.text);
+		} finally {
+			shell.dispose();
```

Comment:
Does this mean that everytime a shell is created, we have to ensure we error catch and ensure it is disposed?

Reply:
You always want to dispose it.  That's what finally is for.  Without a catch, errors will be bubbled as normal.